httpclient5/src/main/java/org/apache/hc/client5/http/impl/nio/HappyEyeballsV2AsyncClientConnectionOperator.java

```diff
@@ -0,0 +1,697 @@
+                                    connectionFutures.forEach(future -> future.cancel(true));
+                                    connectionFuture.complete((ManagedAsyncClientConnection) result);
+                                } else {
+                                    // If there is an exception, complete the connectionFuture exceptionally with the exception
+                                    connectionFuture.completeExceptionally(new ConnectException("Failed to connect to any address for " + host));
+                                }
+                                // Invoke the callback if provided
+                                if (callback != null) {
+                                    connectionFuture.whenComplete((conn, ex) -> {
+                                        if (ex != null) {
+                                            callback.failed(new Exception(ex));
+                                        } else {
+                                            callback.completed(conn);
+                                        }
+                                    });
+                                }
+                                return connectionFuture;
+                            });
+                })
+                .exceptionally(e -> {
+                    connectionFuture.completeExceptionally(e);
+                    if (callback != null) {
+                        callback.failed(new Exception(e));
+                    }
+                    return null;
+                });
+
+        return connectionFuture;
+    }
+
+    /**
+     * Asynchronously resolves the DNS for the given host name and returns a CompletableFuture that will be completed
+     * with an array of InetAddress objects representing the IP addresses of the host.
+     * The resolution of AAAA records is delayed by the configured resolution delay to allow for a chance for A records to be
+     * returned first.
+     *
+     * @param host the host name to resolve DNS for
+     * @return a CompletableFuture that will be completed with an array of InetAddress objects representing the IP addresses
+     */
+    private CompletableFuture<InetAddress[]> resolveDnsAsync(final String host) {
+        final CompletableFuture<InetAddress[]> dnsFuture = new CompletableFuture<>();
+        CompletableFuture.runAsync(() -> {
+            try {
+                final InetAddress[] inetAddresses = dnsResolver.resolve(host);
+                dnsFuture.complete(inetAddresses);
+
+                // Introduce a delay before resolving AAAA records after receiving A records
+                if (inetAddresses.length > 0) {
+                    scheduler.schedule(() -> {
+                        try {
+                            final InetAddress[] inet6Addresses = dnsResolver.resolve(host);
+                            dnsFuture.complete(inet6Addresses);
```

Comment:
You can't complete `dnsFuture` twice. The `complete` call on line 408 will always put `dnsFuture` into a finished state.

Reply:
And please don't use `obtrudeValue` to work around `complete` being one-and-done